Cabal/doc/nix-local-build.markdown

```diff
@@ -0,0 +1,364 @@
+
+~~~~
+FilePathGlob    ::= FilePathRoot FilePathGlobRel
+FilePathRoot    ::= {- empty -}        # relative to cabal.project
+                  | "/"                # Unix root
+                  | [a-zA-Z] ":" [/\\] # Windows root
+                  | "~"                # home directory
+FilePathGlobRel ::= Glob "/"  FilePathGlobRel # Unix directory
+                  | Glob "\\" FilePathGlobRel # Windows directory
+                  | Glob         # file
+                  | {- empty -}  # trailing slash
+Glob      ::= GlobPiece *
+GlobPiece ::= "*"            # wildcard
+            | [^*{},/\\] *   # literal string
+            | "\\" [*{},]    # escaped reserved character
+            | "{" Glob "," ... "," Glob "}" # union (match any of these)
+~~~~
+
+Wildcards and unions have the obvious semantics.  A wildcard like
+`*/foo.cabal` does NOT match `.hidden/foo.cabal`; you'll need to
+explicitly write a period to match this.
+
+## Specifying the local packages ##
+
+The following top-level options specify what the local packages of a
+project are:
+
+`packages:` _package location list_ (space separated, default: `./*.cabal`)
+:   Specifies the list of package locations which contain
+    the local packages to be built by this project.  Package locations
+    can take the following forms:
+
+    1. They can specify a Cabal file, or a directory containing
+       a Cabal file, e.g.,
+       `packages: Cabal cabal-install/cabal-install.cabal`
+
+    2. They can specify a glob-style wildcards, which must match one
+       or more (a) directories containing a (single) Cabal file, (b) Cabal
+       files (extension `.cabal`), or (c) ~~tarballs which
+       contain Cabal packages (extension `.tar.gz`)~~ (not implemented yet).  For example, to
+       match all Cabal files in all subdirectories, as well as the Cabal
+       projects in the parent directories `foo` and `bar`, use
+       `packages: */*.cabal ../{foo,bar}/`
+
+    3. ~~They can specify an `http`, `https` or `file` URL, representing
+       the path to a remote tarball to be downloaded and built.~~ (not
+       implemented yet)
+
+`optional-packages:` _package location list_ (space separated, default: `./*/*.cabal`)
+:   Like `packages:`, specifies a list of package locations containing
+    local packages to be built.  Unlike `packages:`, if we glob for
+    a package, it is permissible for the glob to match against zero
+    packages.  The intended use-case for `optional-packages` is to make
+    it so that vendored packages can be automatically picked up if
+    they are placed in a subdirectory, but not error if there aren't
+    any.
+
+`extra-packages:` _package list with version bounds_ (comma separated)
+:   ~~Specifies a list of external packages from Hackage which should be
+    considered local packages.~~ (Not implemented)
+
+~~There is also a stanza `source-repository-package` for specifying
+packages from an external version control or private Hackage.~~ (Not
```

Comment:
not a private hackage, just scm. You can specify additional hackage repos, but that's a different section, same syntax as in ~/.cabal/config for the global one. Though note there's no new-update so this doesn't work well yet.